frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -3590,6 +3596,20 @@ const IR::Node *TypeInference::postorder(IR::MethodCallExpression *expression) {
                 param = *paramIt;
             }
 
+            if (param->type->is<IR::Type_Dontcare>())
+                typeError(
+                    "%1%: Could not infer a type for parameter %2% "
+                    "(inferred type is don't care '_')",
```

Comment:
Shouldn't this test here be for `arg->type` rather than `param->type`?  `arg->type` is the type from the argument expression (what we're inferring) while `param->type` will be the declared type of the parameter.  As is, this would seem to flag any time we instantiate something a with a don't care and then actually infer something for it, which is very much the opposite of what the message is saying.

Reply:
This is really about the instantiation of a generic extern, where the generic types are used to specify parameters to some of the methods. You are allowed to use _ (don't care) for a type argument when you instantiate an extern, but then you are not supposed to call any of the methods that depend on that type. Here we detect the case when after type unification we still don't know what the parameter type is (it only unifies with _).

Reply:
It should be ok to use `_` as a type argument if type inferencing can infer what the type should be based on the use.  As it is, it will ALWAYS flag an error using `_` as a type argument.